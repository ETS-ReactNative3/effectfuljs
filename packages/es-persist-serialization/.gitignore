index.js
node_modules
tmp-*
*.orig
*.log
*~
*.map
.DS_Store
.tern-port
err-log.json
