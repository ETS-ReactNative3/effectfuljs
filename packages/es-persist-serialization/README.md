# @effectful/es-persist-serialization

The library uses [@effectful/es-persist](https://github.com/awto/effectfuljs/tree/master/packages/serialization)
to store async function state for
[@effectful/es-persist](https://github.com/awto/effectfuljs/tree/master/packages/es-persist).

It re-exports both of the libraries, and `@effectful/serialization` is
imported as peer dependency.

## Usage

Install from npm:

```
$ npm install --save-dev @effectful/es
$ npm install --save @effectful/es-persist-serialization @effectful/serialization
```

In .babelrc:

```json
{
  "presets": ["@effectful/es-persist-serialization/transform"]
}

```

If there are other babel presets, add `"passPerPreset":true`.

## License

Distributed under the terms of the [The MIT License (MIT)](LICENSE). 

