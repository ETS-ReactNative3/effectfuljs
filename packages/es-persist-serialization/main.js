import * as S from "@effectful/serialization";
import * as R from "@effectful/es-persist";
export {
  generator,
  generatorFunction,
  iterator,
  iteratorM,
  forInIterator,
  contextSymbol,
  sendSymbol,
  awaitSymbol,
  context,
  async,
  asyncGenerator,
  lock,
  all,
  any,
  producer,
  current,
  idle,
  abort,
  managed
} from "@effectful/es-persist";

export {
  read,
  write,
  regConstructor,
  regOpaqueObject,
  bind,
  descriptorSymbol,
  parse,
  stringify,
  regDescriptor,
  regOpaquePrim
} from "@effectful/serialization";

R.constructors.forEach(v => S.regConstructor(v));
export const asyncFunction = R.makeAsyncFunctionConstructor(S.regConstructor);
export const asyncGeneratorFunction = R.makeAsyncGeneratorFunctionConstructor(
  S.regConstructor
);
