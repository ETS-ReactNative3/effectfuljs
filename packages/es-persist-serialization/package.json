{
  "name": "@effectful/es-persist-serialization",
  "version": "1.3.6",
  "description": "@effectful/es-persist with @effectful/serialization for storing state",
  "main": "index.js",
  "module": "main.js",
  "jsnext:main": "main.js",
  "scripts": {
    "test": "mocha",
    "build": "babel --config-file ../es/babel.cjs.json -o index.js main.js"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/awto/effectfuljs.git"
  },
  "keywords": [
    "effectful",
    "effectfuljs",
    "monad",
    "linq"
  ],
  "author": "Vitaliy Akimov <vitaliy.akimov@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/awto/effectfuljs/issues"
  },
  "homepage": "https://github.com/awto/effectfuljs#readme",
  "peerDependencies": {
    "@effectful/serialization": "^1.0.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.0.0",
    "@effectful/serialization": "^1.3.6",
    "mocha": "^5.0.5"
  },
  "dependencies": {
    "@effectful/es-persist": "^1.3.1"
  },
  "gitHead": "d3b25d1148c0b8da0bb45a6a52b8b8cac3150ca5"
}
