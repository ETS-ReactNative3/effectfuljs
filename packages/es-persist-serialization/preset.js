module.exports = require("@effectful/es-persist/preset")
  .options({importRT:"@effectful/es-persist-serialization"})
