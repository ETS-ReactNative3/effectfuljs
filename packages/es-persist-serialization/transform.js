module.exports = require("@effectful/es-persist/transform").options({
  importRT: "@effectful/es-persist-serialization"
});
