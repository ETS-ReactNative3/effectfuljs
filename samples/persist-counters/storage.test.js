import * as R from "@effectful/es-persist-serialization"
import saveLocal from "./storage"

const getItem = jest.fn()
const setItem = jest.fn()

global.localStorage = { getItem, setItem }

test("saveLocal", () => {
  let cb, ce
  const exit = { resume() { cb() }, reject(e) { cb(e) }  }
  async function* data() {
    yield {type:"FLUSH"}
  }
  R.regOpaqueObject(exit,"saveLocalExit");
  R.regOpaqueObject(data,"saveLocalIter");
  (async () => {
    if (!await R.managed)
      return
    expect.assertions(3)
    for await(const i of saveLocal(data()))
      expect(R.write(i)).toMatchSnapshot()
    expect(getItem).toHaveBeenLastCalledWith("counters")
    expect(setItem.mock.calls).toMatchSnapshot()
  })()[R.awaitSymbol](exit)
  return new Promise((r,e) => (cb = r, ce = e))
})

