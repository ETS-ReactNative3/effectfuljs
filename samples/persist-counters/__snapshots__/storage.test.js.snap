// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saveLocal 1`] = `
Object {
  "f": Array [
    Array [
      "type",
      "FLUSH",
    ],
  ],
}
`;

exports[`saveLocal 2`] = `
Array [
  Array [
    "counters",
    Object {
      "resume": Object {
        "f": Array [
          Array [
            "running",
            Object {
              "$": "Set",
              "l": Array [
                Object {
                  "r": 3,
                },
                Object {
                  "r": 0,
                },
                Object {
                  "r": 2,
                },
                Object {
                  "r": 4,
                },
              ],
            },
          ],
          Array [
            "resume",
            Object {
              "r": 4,
            },
          ],
        ],
        "x": Array [
          Object {
            "$": "_data",
            "f": Array [
              Array [
                "queue",
                Array [],
              ],
              Array [
                "$step",
                6,
              ],
              Array [
                "$last",
                5,
              ],
              Array [
                "$arg",
                Object {
                  "$": "undefined",
                },
              ],
            ],
          },
          Object {
            "f": Array [
              Array [
                "type",
                "FLUSH",
              ],
            ],
          },
          Object {
            "$": "saveLocal",
            "f": Array [
              Array [
                "queue",
                Array [
                  Object {
                    "$": "Residual",
                    "f": Array [
                      Array [
                        "dest",
                        Array [
                          Object {
                            "r": 3,
                          },
                        ],
                      ],
                      Array [
                        "state",
                        0,
                      ],
                    ],
                  },
                ],
              ],
              Array [
                "_input",
                null,
              ],
              Array [
                "$step",
                13,
              ],
              Array [
                "$last",
                12,
              ],
              Array [
                "$arg",
                Object {
                  "$": "undefined",
                },
              ],
              Array [
                "_loop",
                Object {
                  "r": 0,
                },
              ],
              Array [
                "_i",
                Object {
                  "r": 1,
                },
              ],
            ],
          },
          Object {
            "$": "Async$1",
            "f": Array [
              Array [
                "dest",
                Array [],
              ],
              Array [
                "state",
                0,
              ],
              Array [
                "value",
                Object {
                  "$": "undefined",
                },
              ],
              Array [
                "_f",
                Object {
                  "f": Array [
                    Array [
                      "_data",
                      Object {
                        "$": "saveLocalIter",
                      },
                    ],
                  ],
                },
              ],
              Array [
                "$step",
                8,
              ],
              Array [
                "cont",
                Object {
                  "$": "Residual",
                  "f": Array [
                    Array [
                      "dest",
                      Array [
                        Object {
                          "$": "saveLocalExit",
                        },
                      ],
                    ],
                    Array [
                      "state",
                      0,
                    ],
                  ],
                },
              ],
              Array [
                "$last",
                7,
              ],
              Array [
                "_loop",
                Object {
                  "r": 2,
                },
              ],
              Array [
                "$arg",
                Object {
                  "$": "undefined",
                },
              ],
              Array [
                "_i",
                Object {
                  "r": 1,
                },
              ],
            ],
          },
          Object {
            "$": "save",
            "f": Array [
              Array [
                "dest",
                Array [],
              ],
              Array [
                "state",
                0,
              ],
              Array [
                "value",
                Object {
                  "$": "undefined",
                },
              ],
              Array [
                "$step",
                6,
              ],
              Array [
                "cont",
                Object {
                  "$": "Residual",
                  "f": Array [
                    Array [
                      "dest",
                      Array [
                        Object {
                          "r": 2,
                        },
                      ],
                    ],
                    Array [
                      "state",
                      0,
                    ],
                  ],
                },
              ],
              Array [
                "$last",
                6,
              ],
              Array [
                "$arg",
                Object {
                  "$": "undefined",
                },
              ],
            ],
          },
        ],
      },
    },
  ],
]
`;
